pkg/agent/cache/manager.go

```diff
@@ -0,0 +1,382 @@
+		go m.expiredCacheEntryHandler(5*time.Second, &wg)
+
+		wg.Add(1)
+		go m.regEntriesHandler(&wg)
+		m.log.Debug("Initializing Cache Manager")
+		m.regEntriesCh <- m.baseRegEntries
+		for {
+
+			var svid []byte
+			var key *ecdsa.PrivateKey
+			select {
+			case reqs := <-m.entryRequestCh:
+				for parentId, entryRequests := range reqs {
+					wg.Add(1)
+					if _, ok := m.spiffeIdEntryMap[parentId]; ok {
+						svid = m.spiffeIdEntryMap[parentId].SVID.SvidCert
+						key = m.spiffeIdEntryMap[parentId].PrivateKey
+					}
+					if parentId == m.baseSPIFFEID {
+						svid = m.baseSVID
+						key = m.baseSVIDKey
+					}
+					conn, err := m.getGRPCConn(svid, key)
+					if err != nil {
+						m.Shutdown(err)
+						break
+					}
+					go m.fetchSVID(entryRequests, node.NewNodeClient(conn), &wg)
+				}
+
+			case newCacheEntry := <-m.cacheEntryCh:
+				m.managedCache.SetEntry(newCacheEntry)
+				m.log.Debug("Updated Cache", "Cache:", newCacheEntry)
+
+			case <-m.ctx.Done():
+				m.log.Debug("Stopping Cache manager")
+				wg.Wait()
+				return
+			}
+		}
+	}()
+}
+
+func (m *manager) fetchSVID(requests []EntryRequest, nodeClient node.NodeClient, wg *sync.WaitGroup) {
+	defer wg.Done()
+	stream, err := nodeClient.FetchSVID(context.Background())
+	if err != nil {
+		m.Shutdown(err)
+		return
+	}
+	for _, req := range requests {
+
+		err := stream.Send(&node.FetchSVIDRequest{Csrs: append([][]byte{}, req.CSR)})
+
+		if err != nil {
+			m.Shutdown(err)
+			return
+		}
+
+		resp, err := stream.Recv()
+		if err != nil {
+			m.Shutdown(err)
+			return
+		}
+		svid := resp.SvidUpdate.Svids[req.entry.RegistrationEntry.SpiffeId]
+		cert, err := x509.ParseCertificate(svid.SvidCert)
+		if err != nil {
+			stream.CloseSend()
+			m.Shutdown(err)
+			return
+		}
+
+		m.regEntriesCh <- resp.SvidUpdate.RegistrationEntries
+		req.entry.SVID = svid
+		req.entry.Expiry = cert.NotAfter
+		select {
+		case m.cacheEntryCh <- req.entry:
+		case <-m.ctx.Done():
+			m.Shutdown(m.ctx.Err())
+			return
+		}
+
+	}
+	if err := stream.CloseSend(); err != nil {
+		m.Shutdown(err)
+		return
+	}
+}
+
+func (m *manager) fetchWithEmptyCSR(svid []byte, key *ecdsa.PrivateKey) {
+	conn, err := m.getGRPCConn(svid, key)
+	if err != nil {
+		m.Shutdown(err)
+		return
+	}
+	stream, err := node.NewNodeClient(conn).FetchSVID(m.ctx)
+	if err != nil {
+		m.Shutdown(err)
+		return
+	}
+	err = stream.Send(&node.FetchSVIDRequest{})
+	if err != nil {
+		m.Shutdown(err)
+		return
+	}
+	strmch := make(chan struct{})
+	go func() {
+		for {
+			resp, err := stream.Recv()
+			if err == io.EOF {
+				close(strmch)
+				stream.CloseSend()
+				return
+			}
+
+			if err != nil {
+				m.Shutdown(err)
+				stream.CloseSend()
+				close(strmch)
+				return
+			}
+
+			select {
+			case m.regEntriesCh <- resp.SvidUpdate.RegistrationEntries:
+			case <-m.ctx.Done():
+				m.Shutdown(m.ctx.Err())
+				stream.CloseSend()
+				close(strmch)
+				return
+			}
+			m.log.Debug(resp)
+		}
+	}()
+	stream.CloseSend()
```

Comment:
curious: what does `CloseSend()` get you over `Close()`?  Using a single `defer stream.Close()` is more readable to me but maybe I'm missing something?

Reply:
the GRPC `ClientStream` interface used here for bi-directional stream only support `CloseSend()` which closes the send side so if there are more messages from the other direction it would still receive it.

Reply:
I see, thanks.  Maybe just `defer stream.CloseSend()` then?

Reply:
calling defer `stream.CloseSend()` would block the stream from closing and creates a deadlock..the go func would never terminate

Reply:
Not sure I understand, but ok.

As of now it's doing

```go
// ...
go func(){
  // ...
  stream.CloseSend()
  // ...
}()
stream.CloseSend()
// ...
```

which doesn't make much sense to me.

Reply:
so `stream.Recv() returns  `err=io.EOF` to signal the end of msgs on the stream which would never happen without sending `stream.ConseSend()` In our `go func` we wait till we receive io.EOF so with defer we would get blocked waiting on `io.EOF` 
Also I removed the `stream.CloseSend()` inside the go func()